syntax = "proto3";

package book_storage;

service BookStorage {
    rpc Write(WriteRequest) returns (WriteResponse);
    rpc Read(ReadRequest) returns (ReadResponse);
    rpc CleanVersion(CleanVersionRequest) returns (CleanVersionResponse);
    rpc CheckStock(CheckStockRequest) returns (CheckStockResponse);
    rpc DecrementStock(StockUpdateRequest) returns (StockUpdateResponse);
    rpc IncrementStock(StockUpdateRequest) returns (StockUpdateResponse);
    rpc Prepare(PrepareRequest) returns (PrepareResponse);
    rpc Commit(CommitRequest) returns (CommitResponse);
    rpc Abort(AbortRequest) returns (AbortResponse);
}

message Book {
    string id = 1;
    string title = 2;
    string author = 3;
    string description = 4;
    int32 copies = 5;
    int32 copiesAvailable = 6;
    string category = 7;
    string img = 8;
    float price = 9;
}

message WriteRequest {
    string key = 1;
    Book value = 2;
}

message WriteResponse {
    bool success = 1;
}

message ReadRequest {
    string key = 1;
}

message ReadResponse {
    Book value = 1;
    bool clean = 2;
    bool found = 3;
}

message CleanVersionRequest {
    string key = 1;
    int32 version = 2;
}

message CleanVersionResponse {
    bool success = 1;
}



message CheckStockRequest {
    string key = 1;
}

message CheckStockResponse {
    int32 copiesAvailable = 1;
    bool success = 2;
    string message = 3;
}

message StockUpdateRequest {
    string key = 1;
    int32 value = 2;
}

message StockUpdateResponse {
    bool success = 1;
}





message PrepareRequest {
  string order_id = 1;
  string order_details_json = 2;
  string message = 3;
}

message PrepareResponse {
  bool willCommit = 1;
  string message = 2;
}

message CommitRequest {
  string order_id = 1;
  string message = 2;
}

message CommitResponse {
  bool success = 1;
  string message = 2;
  string order_details_json = 3;
}

message AbortRequest {
  string order_id = 1;
  string message = 2;
}

message AbortResponse {
  bool success = 1;
  string message = 2;
  string order_details_json = 3;
}